Wrapping text dataset in geo dataset fails, other way around not

### Version

4.6.0 latest SNAPSHOT

### What happened?

When using both, the text and the geo index the order of wrapping both matters.

**Works**: TDB2 -> Geo -> Text -> Service
**Fails**:  TDB2 -> Text  -> Geo-> Service


The reason is basically that the geo index while preparing calls
```
boolean isEmpty = dataset.calculate(()->dataset.isEmpty());
```
which as far as I understand is a "promoting" transaction. The problem is, that the text index implementation never implemented handling those, see the following lines in [DatasetGraphText](https://github.com/apache/jena/blob/main/jena-text/src/main/java/org/apache/jena/query/text/DatasetGraphText.java#L159-L161)

The question is now:
- Do we have to implement handling that transaction type?
- Or should we call `dataset.calc()` instead which is no promoting type? What would be the implications here?

Attached are 2 assembler files, need GeoSPARQL module enabled in Fuseki (we should make this default anyways I think and add it to the POM similar to text being enabled by default)

[fails_tdb2_text_geo_assembler.ttl.gz](https://github.com/apache/jena/files/9118293/fails_tdb2_text_geo_assembler.ttl.gz)
[works_tdb2_geo_text_assembler.ttl.gz](https://github.com/apache/jena/files/9118295/works_tdb2_geo_text_assembler.ttl.gz)
 

### Relevant output and stacktrace

```shell
Caused by: java.lang.UnsupportedOperationException: begin(READ_PROMOTE)
	at org.apache.jena.query.text.DatasetGraphText.begin(DatasetGraphText.java:161) ~[fuseki-server.jar:4.6.0-SNAPSHOT]
	at org.apache.jena.sparql.core.DatasetImpl.begin(DatasetImpl.java:114) ~[fuseki-server.jar:4.6.0-SNAPSHOT]
	at org.apache.jena.system.Txn.calc(Txn.java:96) ~[fuseki-server.jar:4.6.0-SNAPSHOT]
	at org.apache.jena.system.Txn.calculate(Txn.java:67) ~[fuseki-server.jar:4.6.0-SNAPSHOT]
	at org.apache.jena.sparql.core.Transactional.calculate(Transactional.java:248) ~[fuseki-server.jar:4.6.0-SNAPSHOT]
	at org.apache.jena.geosparql.assembler.GeoAssembler.prepareSpatialExtension(GeoAssembler.java:176) ~[fuseki-server.jar:4.6.0-SNAPSHOT]
	at org.apache.jena.geosparql.assembler.GeoAssembler.createDataset(GeoAssembler.java:158) ~[fuseki-server.jar:4.6.0-SNAPSHOT]
	at org.apache.jena.sparql.core.assembler.DatasetAssembler.open(DatasetAssembler.java:40) ~[fuseki-server.jar:4.6.0-SNAPSHOT]
	at org.apache.jena.sparql.core.assembler.DatasetAssembler.open(DatasetAssembler.java:33) ~[fuseki-server.jar:4.6.0-SNAPSHOT]
	at org.apache.jena.assembler.assemblers.AssemblerGroup$PlainAssemblerGroup.openBySpecificType(AssemblerGroup.java:157) ~[fuseki-server.jar:4.6.0-SNAPSHOT]
```


### Are you interested in making a pull request?

_No response_

--- a/jena-geosparql/src/main/java/org/apache/jena/geosparql/assembler/GeoAssembler.java
+++ b/jena-geosparql/src/main/java/org/apache/jena/geosparql/assembler/GeoAssembler.java
@@ -166,7 +166,7 @@ public class GeoAssembler extends DatasetAssembler {
     }
 
     private static void prepareSpatialExtension(Dataset dataset, String spatialIndex){
-        boolean isEmpty = dataset.calculate(()->dataset.isEmpty());
+        boolean isEmpty = dataset.calculateRead(()->dataset.isEmpty());
         if ( isEmpty && spatialIndex != null ) {
             LOG.warn("Dataset empty. Spatial Index not constructed. Server will require restarting after adding data and any updates to build Spatial Index.");
             return;

--- a/jena-integration-tests/src/test/files/GeoAssembler/fuseki-geo-text-tdb2.ttl
+++ b/jena-integration-tests/src/test/files/GeoAssembler/fuseki-geo-text-tdb2.ttl
@@ -1,0 +1,68 @@
+PREFIX : <https://data.coypu.org/>
+PREFIX fuseki:    <http://jena.apache.org/fuseki#>
+PREFIX rdf:       <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
+PREFIX rdfs:      <http://www.w3.org/2000/01/rdf-schema#>
+PREFIX tdb1:      <http://jena.hpl.hp.com/2008/tdb#>
+PREFIX tdb2:      <http://jena.apache.org/2016/tdb#>
+PREFIX text:  <http://jena.apache.org/text#>
+PREFIX ja:        <http://jena.hpl.hp.com/2005/11/Assembler#>
+PREFIX geosparql: <http://jena.apache.org/geosparql#>
+PREFIX ex:        <http://www.example.org/resources#>
+
+# Text and Geo service 
+<#serviceTest> rdf:type fuseki:Service;
+    fuseki:name "ds";
+    fuseki:endpoint [ fuseki:operation fuseki:query ; ] ;
+    fuseki:dataset <#testGeoDS> .
+
+# Text DS
+<#testTextDS> rdf:type text:TextDataset ;
+    text:dataset   <#testDS> ;
+    text:index     <#testIndexLucene> ;
+    .
+
+# Text index description
+<#testIndexLucene> rdf:type text:TextIndexLucene ;
+    text:directory "mem" ;
+    text:entityMap <#testEntMap> ;
+    text:storeValues true ; 
+    text:analyzer [ a text:StandardAnalyzer ] ;
+    text:queryParser text:AnalyzingQueryParser ;
+    text:multilingualSupport true ; # optional
+    .
+
+<#testEntMap> rdf:type text:EntityMap ;
+    text:defaultField     "label" ;
+    text:entityField      "uri" ;
+    text:uidField         "uid" ;
+    text:langField        "lang" ;
+    text:graphField       "graph" ;
+    text:map (
+        [ text:field "label" ; 
+          text:predicate rdfs:label ]      
+    ) ;
+    .
+
+<#testGeoDS> rdf:type geosparql:geosparqlDataset ;
+    ## If absent build an in-memory index each time.
+    ## geosparql:spatialIndexFile     "";
+
+    ## Default settings. See documentation for meanings.
+    geosparql:inference            false ;
+    geosparql:queryRewrite         true ;
+    geosparql:indexEnabled         true ;
+    geosparql:applyDefaultGeometry false ;
+
+    # 3 item lists: [Geometry Literal, Geometry Transform, Query Rewrite]
+    geosparql:indexSizes           "-1,-1,-1" ;       # Default - unlimited.
+    geosparql:indexExpires         "5000,5000,5000" ; # Default - time in milliseconds.
+
+    ## Required - data over which GeoSPARQL is applied.
+    geosparql:dataset <#testTextDS> ;
+    .
+
+
+# TDB2 dataset
+<#testDS> rdf:type tdb2:DatasetTDB2 ;
+    tdb2:location "--mem--" ;
+    .

--- a/jena-integration-tests/src/test/files/GeoAssembler/fuseki-text-geo-tdb2.ttl
+++ b/jena-integration-tests/src/test/files/GeoAssembler/fuseki-text-geo-tdb2.ttl
@@ -1,0 +1,69 @@
+PREFIX : <https://data.coypu.org/>
+PREFIX fuseki:    <http://jena.apache.org/fuseki#>
+PREFIX rdf:       <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
+PREFIX rdfs:      <http://www.w3.org/2000/01/rdf-schema#>
+PREFIX tdb1:      <http://jena.hpl.hp.com/2008/tdb#>
+PREFIX tdb2:      <http://jena.apache.org/2016/tdb#>
+PREFIX text:  <http://jena.apache.org/text#>
+PREFIX ja:        <http://jena.hpl.hp.com/2005/11/Assembler#>
+PREFIX geosparql: <http://jena.apache.org/geosparql#>
+PREFIX ex:        <http://www.example.org/resources#>
+
+# Text and Geo service 
+<#serviceTest> rdf:type fuseki:Service;
+    fuseki:name "ds";
+    fuseki:endpoint [ fuseki:operation fuseki:query ; ] ;
+    fuseki:dataset <#testTextDS> .
+
+# Text DS
+<#testTextDS> rdf:type text:TextDataset ;
+    text:dataset   <#testGeoDS> ;
+    text:index     <#testIndexLucene> ;
+    .
+
+# Text index description
+<#testIndexLucene> rdf:type text:TextIndexLucene ;
+    text:directory "mem" ;
+    text:entityMap <#testEntMap> ;
+    text:storeValues true ; 
+    text:analyzer [ a text:StandardAnalyzer ] ;
+    text:queryParser text:AnalyzingQueryParser ;
+    text:multilingualSupport true ; # optional
+    .
+
+<#testEntMap> rdf:type text:EntityMap ;
+    text:defaultField     "label" ;
+    text:entityField      "uri" ;
+    text:uidField         "uid" ;
+    text:langField        "lang" ;
+    text:graphField       "graph" ;
+    text:map (
+        [ text:field "label" ; 
+          text:predicate rdfs:label ]      
+    ) ;
+    .
+
+<#testGeoDS> rdf:type geosparql:geosparqlDataset ;
+    ## If absent build an in-memory index each time.
+    ## geosparql:spatialIndexFile     "target/tdb2/test/spatial.index";
+
+    ## Default settings. See documentation for meanings.
+    geosparql:inference            false ;
+    geosparql:queryRewrite         true ;
+    geosparql:indexEnabled         true ;
+    geosparql:applyDefaultGeometry false ;
+
+    # 3 item lists: [Geometry Literal, Geometry Transform, Query Rewrite]
+    geosparql:indexSizes           "-1,-1,-1" ;       # Default - unlimited.
+    geosparql:indexExpires         "5000,5000,5000" ; # Default - time in milliseconds.
+
+    ## Required - data over which GeoSPARQL is applied.
+    geosparql:dataset <#testDS> ;
+    .
+
+
+# TDB2 dataset
+<#testDS> rdf:type tdb2:DatasetTDB2 ;
+    tdb2:location "target/tdb2/test" ;
+    .
+

--- a/jena-integration-tests/src/test/java/org/apache/jena/geosparql/TestGeoAssembler.java
+++ b/jena-integration-tests/src/test/java/org/apache/jena/geosparql/TestGeoAssembler.java
@@ -20,8 +20,11 @@ package org.apache.jena.geosparql;
 
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assume.assumeFalse;
 
 import org.apache.jena.atlas.lib.FileOps;
+import org.apache.jena.base.Sys;
 import org.apache.jena.fuseki.main.FusekiServer;
 import org.apache.jena.geosparql.assembler.VocabGeoSPARQL;
 import org.apache.jena.query.Dataset;
@@ -57,7 +60,6 @@ public class TestGeoAssembler {
         ds.executeWrite(()->{
             RDFDataMgr.read(ds, DIR+"geosparql_test.rdf");
         });
-
     }
 
     @Test public void geoAssemblerTest() {
@@ -66,18 +68,29 @@ public class TestGeoAssembler {
     }
 
     @Test public void testBasicFusekiGeoAssembler() {
-        test("geo-config.ttl", "ds1");
+        testBuildDataQuery("geo-config.ttl", "ds1");
     }
 
     @Test public void testExampleFusekiGeoAssembler() {
-        test("geo-config-ex.ttl", "ds2");
+        testBuildDataQuery("geo-config-ex.ttl", "ds2");
     }
 
     @Test public void testMemFusekiGeoAssembler() {
-        test("geo-config-mem.ttl", "ds3");
+        testBuildDataQuery("geo-config-mem.ttl", "ds3");
     }
 
-    private void test(String filename, String dbName) {
+    @Test public void testFusekiTextGeoTDB2() {
+        assumeFalse(Sys.isWindows);
+        testBuildPing("fuseki-text-geo-tdb2.ttl", "ds");
+    }
+
+    @Test public void testFusekiGeoTextTDB2() {
+        assumeFalse(Sys.isWindows);
+        testBuildPing("fuseki-geo-text-tdb2.ttl", "ds");
+    }
+
+    // Test for the configurations with data loaded in the assembler
+    private void testBuildDataQuery(String filename, String dbName) {
         FusekiServer server = FusekiServer.create().port(0)
             .parseConfigFile("file:"+DIR+filename)
             .build();
@@ -99,6 +112,22 @@ public class TestGeoAssembler {
                 RowSet rs = QueryExec.service(URL).query(queryStr).select();
                 long x = RowSetOps.count(rs);
                 assertEquals(8, x);
+        } finally {
+            server.stop();
+        }
+    }
+
+    // Test for valid configurations.
+    private void testBuildPing(String config, String dbName) {
+        FusekiServer server = FusekiServer.create().port(0)
+                .parseConfigFile("file:"+DIR+config)
+                .build();
+        try {
+            server.start();
+            int port = server.getPort();
+            String URL = "http://localhost:"+port+"/"+dbName;
+            boolean ask = QueryExec.service(URL).query("ASK{}").ask();
+            assertTrue(ask);
         } finally {
             server.stop();
         }

